### PR TITLE
Sheet visibility - XLSX

### LINKED ISSUE #445: Sheet Visibility
Hello, I did changes to allow to retrieve and set the sheet visibility and i have seen that i need to open an issue first to make a pull request. Can I?

### PR DESCRIPTION
**Pull request by alexisparron**
Fixes #445 
I did changes to allow to retrieve and set the sheet visibility

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @alexisparron has signed the CLA. Thanks for the pull request!

File: src/Spout/Writer/XLSX/Helper/FileSystemHelper.php [hunk 1]
```diff
@@ -289,8 +289,10 @@ public function createWorkbookFile($worksheets)
         /** @var Worksheet $worksheet */
         foreach ($worksheets as $worksheet) {
             $worksheetName = $worksheet->getExternalSheet()->getName();
+            $worksheetState = ( $worksheet->getExternalSheet()->isVisible() ) ? 'visible' : 'hidden';
+
             $worksheetId = $worksheet->getId();
-            $workbookXmlFileContents .= '<sheet name="' . $escaper->escape($worksheetName) . '" sheetId="' . $worksheetId . '" r:id="rIdSheet' . $worksheetId . '"/>';
+            $workbookXmlFileContents .= '<sheet name="' . $escaper->escape($worksheetName) . '" sheetId="' . $worksheetId . '" state="' . $worksheetState. '" r:id="rIdSheet' . $worksheetId . '"/>';
```
Comment (adrilo) [inline src/Spout/Writer/XLSX/Helper/FileSystemHelper.php]: `state` or `status`? When you read the data, you search for `status`

File: src/Spout/Writer/Common/Sheet.php [hunk 2]
```diff
@@ -33,6 +33,9 @@ class Sheet
     /** @var string Name of the sheet */
     protected $name;
 
+     /** @var bool isVisible visibility of the sheet */
```
Comment (adrilo) [inline src/Spout/Writer/Common/Sheet.php]: `/** @var bool Visibility of the sheet */` (removing the reference to the var name)

File: src/Spout/Writer/Common/Sheet.php [hunk 3]
```diff
@@ -71,6 +75,23 @@ public function getName()
...
     /**
+     * @api
+     * @return bool isVisible visibility of the sheet
+     */
+    public function isVisible()
+    {
+        return $this->isVisible;
+    }
+
+    public function setVisibility($visibility = true)
```
Comment (adrilo) [inline src/Spout/Writer/Common/Sheet.php]: Add PHPDoc. Change the signature to setIsVisible($isVisible) - name change and no default param.

File: src/Spout/Writer/Common/Sheet.php [hunk 4]
```diff
@@ -50,6 +53,7 @@ public function __construct($sheetIndex, $associatedWorkbookId)
 
         $this->stringHelper = new StringHelper();
         $this->setName(self::DEFAULT_SHEET_NAME_PREFIX . ($sheetIndex + 1));
+        $this->setVisibility();
```
Comment (adrilo) [inline src/Spout/Writer/Common/Sheet.php]: $this->setIsVisible(true);

File: src/Spout/Reader/XLSX/Sheet.php [hunk 5]
```diff
@@ -76,4 +80,13 @@ public function isActive()
     {
         return $this->isActive;
     }
+
+     /**
+     * @api
+     * @return bool Whether the sheet was visible or not
```
Comment (adrilo) [inline src/Spout/Reader/XLSX/Sheet.php]: the sheet *is* visible (was -> is)

File: src/Spout/Reader/XLSX/Sheet.php [hunk 6]
```diff
@@ -24,6 +24,9 @@ class Sheet implements SheetInterface
     /** @var bool Whether the sheet was the active one */
     protected $isActive;
 
+    /** @var bool Whether the sheet was visible or not */
```
Comment (adrilo) [inline src/Spout/Reader/XLSX/Sheet.php]: the sheet *is* visible (was -> is)

File: src/Spout/Reader/XLSX/Helper/SheetHelper.php [hunk 7]
```diff
@@ -105,16 +106,18 @@ protected function getSheetFromSheetXMLNode($xmlReaderOnSheetNode, $sheetIndexZe
     {
         $sheetId = $xmlReaderOnSheetNode->getAttribute(self::XML_ATTRIBUTE_R_ID);
         $escapedSheetName = $xmlReaderOnSheetNode->getAttribute(self::XML_ATTRIBUTE_NAME);
+        $sheetStatus = $xmlReaderOnSheetNode->getAttribute(self::XML_ATTRIBUTE_STATUS);
 
         /** @noinspection PhpUnnecessaryFullyQualifiedNameInspection */
         $escaper = \Box\Spout\Common\Escaper\XLSX::getInstance();
         $sheetName = $escaper->unescape($escapedSheetName);
+        $sheetVisible = ($sheetStatus == 'visible') ? true : false;
```
Comment (adrilo) [inline src/Spout/Reader/XLSX/Helper/SheetHelper.php]: Replace 'visible' by const.
And then `$isSheetVisible = ($sheetStatus === self::SHEET_STATUS_VISIBLE);`

**Comment by adrilo** [comment]
@alexisparron Thanks for the PR! I left some comments, please take a look.
There are 2 problems though:
1. there are no tests
2. this introduces a discrepancy between the XLSX and ODS reader/writer.

For the tests, I'll leave that to you. For the discrepancy, I figure we can't ask contributors to know about both XLSX and ODS formats. However, we try hard to have the same features/api for XLSX and ODS readers/writers so before merging your change, I'll need to make sure that adding this behavior to the ODS reader/writer is doable and a relatively simple change so that it can be done quickly after this change.

**Comment by alexisparron** [comment]
Thanks @adrilo i´m with the tests.

File: src/Spout/Writer/XLSX/Helper/FileSystemHelper.php (hunk 1 above)
Comment (alexisparron) [inline src/Spout/Writer/XLSX/Helper/FileSystemHelper.php]: Is state, I dont know why it was working correctly before.

File: src/Spout/Reader/XLSX/Helper/SheetHelper.php [hunk 8]
```diff
@@ -27,8 +27,11 @@ class SheetHelper
     const XML_ATTRIBUTE_ACTIVE_TAB = 'activeTab';
     const XML_ATTRIBUTE_R_ID = 'r:id';
     const XML_ATTRIBUTE_NAME = 'name';
+    const XML_ATTRIBUTE_STATE = 'state';
     const XML_ATTRIBUTE_ID = 'Id';
     const XML_ATTRIBUTE_TARGET = 'Target';
+    
+    const SHEET_STATUS_VISIBLE = 'visible';
```
Comment (adrilo) [inline src/Spout/Reader/XLSX/Helper/SheetHelper.php]: SHEET_STATE_VISIBLE

File: src/Spout/Reader/XLSX/Helper/SheetHelper.php [hunk 9]
```diff
@@ -105,16 +108,18 @@ protected function getSheetFromSheetXMLNode($xmlReaderOnSheetNode, $sheetIndexZe
     {
         $sheetId = $xmlReaderOnSheetNode->getAttribute(self::XML_ATTRIBUTE_R_ID);
         $escapedSheetName = $xmlReaderOnSheetNode->getAttribute(self::XML_ATTRIBUTE_NAME);
+        $sheetState = $xmlReaderOnSheetNode->getAttribute(self::XML_ATTRIBUTE_STATE);
 
         /** @noinspection PhpUnnecessaryFullyQualifiedNameInspection */
         $escaper = \Box\Spout\Common\Escaper\XLSX::getInstance();
         $sheetName = $escaper->unescape($escapedSheetName);
+        $sheetVisible = ($sheetState === self::SHEET_STATUS_VISIBLE) ? true : false;
```
Comment (adrilo) [inline src/Spout/Reader/XLSX/Helper/SheetHelper.php]: you can remove the `? true : false` as this is redundant. `$sheetVisible = ($sheetState === self::SHEET_STATUS_VISIBLE);` is enough

Comment (adrilo) [inline src/Spout/Reader/XLSX/Helper/SheetHelper.php]: Also, please rename `$sheetVisible` to `$isSheetVisible` (naming convention for bool)

File: src/Spout/Writer/Common/Sheet.php [hunk 10]
```diff
@@ -71,6 +75,28 @@ public function getName()
...
+    {
+        return $this->isVisible;
+    }
+
+    /**
+     * @api
+     * @param string $visibility Visibility of the sheet
+     * @return Sheet
+     */
+    public function setIsVisible($visibility = true)
```
Comment (adrilo) [inline src/Spout/Writer/Common/Sheet.php]: same here. please rename `$visibility` to `$isVisible`

File: tests/Spout/Writer/XLSX/SheetTest.php [hunk 11]
```diff
@@ -76,6 +77,18 @@ public function testSetSheetNameShouldThrowWhenNameIsAlreadyUsed()
     }
 
     /**
+     * @return void
+     */
+    public function testSetSheetVisibilityShouldCreateSheetHidden()
+    {
+        $fileName = 'test_set_visibility_should_create_sheet_hidden.xlsx';
+        // $customSheetName = 'CustomName';
```
Comment (adrilo) [inline tests/Spout/Writer/XLSX/SheetTest.php]: remove commented line

File: tests/Spout/Writer/XLSX/SheetTest.php [hunk 12]
```diff
@@ -118,6 +131,25 @@ private function writeDataToMulitpleSheetsAndReturnSheets($fileName)
     }
 
     /**
+     * @param string $fileName
+     * @return Sheet
+     */
+    private function writeDataAndReturnSheetHidden($fileName)
```
Comment (adrilo) [inline tests/Spout/Writer/XLSX/SheetTest.php]: rename writeDataToHiddenSheet

File: tests/Spout/Writer/XLSX/SheetTest.php [hunk 13]
```diff
@@ -118,6 +131,25 @@ private function writeDataToMulitpleSheetsAndReturnSheets($fileName)
     }
 
     /**
+     * @param string $fileName
+     * @return Sheet
```
Comment (adrilo) [inline tests/Spout/Writer/XLSX/SheetTest.php]: `@return void`

File: tests/Spout/Writer/XLSX/SheetTest.php [hunk 14]
```diff
@@ -76,6 +77,18 @@ public function testSetSheetNameShouldThrowWhenNameIsAlreadyUsed()
...
+    public function testSetSheetVisibilityShouldCreateSheetHidden()
+    {
+        $fileName = 'test_set_visibility_should_create_sheet_hidden.xlsx';
+        // $customSheetName = 'CustomName';
+        $this->writeDataAndReturnSheetHidden($fileName);
+
+        $this->assertSheetVisibilityEquals(false,  $fileName, "The sheet visibility should have been changed to 'false'");
+    }
+
+    /**
```
Comment (adrilo) [inline tests/Spout/Writer/XLSX/SheetTest.php]: just below, can you rename the function `writeDataAndReturnSheetWithCustomName` into `writeDataToSheetWithCustomName` and change the `@return Sheet` to `@return void` please ?

File: tests/Spout/Writer/XLSX/SheetTest.php [hunk 15]
```diff
@@ -131,4 +163,19 @@ private function assertSheetNameEquals($expectedName, $fileName, $message = '')
...
         $this->assertContains("<sheet name=\"$expectedName\"", $xmlContents, $message);
     }
+
+    /**
+     * @param string $expectedVisibility
+     * @param string $fileName
+     * @param string $message
+     * @return void
+     */
+    private function assertSheetVisibilityEquals($expectedVisibility, $fileName, $message = '')
```
Comment (adrilo) [inline tests/Spout/Writer/XLSX/SheetTest.php]: This function is confusing... First it does not use `$expectedVisibility`. Also, it does a global search across the workbook.xml file. This means that if a file has 2 sheets, one hidden, one visible, this function will always return `true`.

File: tests/Spout/Writer/XLSX/SheetTest.php [hunk 16]
```diff
@@ -131,4 +162,18 @@ private function assertSheetNameEquals($expectedName, $fileName, $message = '')
 
         $this->assertContains("<sheet name=\"$expectedName\"", $xmlContents, $message);
     }
+
+    /**
+     * @param string $fileName
+     * @param string $message
+     * @return void
+     */
+    private function assertIsHiddenSheet( $fileName, $message = '')
```
Comment (adrilo) [inline tests/Spout/Writer/XLSX/SheetTest.php]: rename to `assertContainsHiddenSheet` + remove space between `(` and `$fileName`

File: tests/Spout/Writer/XLSX/SheetTest.php [hunk 17]
```diff
@@ -76,11 +77,22 @@ public function testSetSheetNameShouldThrowWhenNameIsAlreadyUsed()
...
 
     /**
+     * @return void
+     */
+    public function testSetSheetVisibilityShouldCreateSheetHidden()
+    {
+        $fileName = 'test_set_visibility_should_create_sheet_hidden.xlsx';
+        $this->writeDataToHiddenSheet($fileName);
+
+        $this->assertIsHiddenSheet( $fileName, "The sheet visibility should have been changed to 'hidden'");
```
Comment (adrilo) [inline tests/Spout/Writer/XLSX/SheetTest.php]: rename to `assertContainsHiddenSheet` + remove space between `(` and `$fileName`

**Comment by adrilo** [comment]
Looking good! Thanks @alexisparron !!
I left 2 very minor comments. Feel free to fix and rebase :) I still need to investigate the feasibility of this on the ODS side. Hopefully soon !

**Comment by adrilo** [comment]
BTW, I investigated the feasibility with ODS files and it seems like it won't be too hard to implement ! So I'll merge your changes as soon as you have fixed the few comments remaining and I'll tackle ODS support.

**Comment by adrilo** [comment]
@alexisparron I took your changes and merged them into the develop_3.0 branch, as we're trying to stop adding features to the 2.x branches. You can see it here: #496 

Thanks for your contribution!

**Pull request closed without merging**